src/engine/input/InputManager.cs

```diff
@@ -0,0 +1,108 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Reflection;
+using Godot;
+
+public class InputManager : Node
+{
+    public override void _Ready()
+    {
+        RecalculateAttributes();
+        PauseMode = PauseModeEnum.Process;
+    }
+
+    public override void _UnhandledInput(InputEvent @event)
+    {
+        RunOnInputAttribute.AttributesWithMethods.ForEach(p => p.Item2.InputReceiver.CheckInput(@event));
+    }
+
+    public override void _Process(float delta)
+    {
+        var disposed = new List<object>();
+        RunOnInputAttribute.AttributesWithMethods.ForEach(p =>
+        {
+            var inputReceiver = p.Item2.InputReceiver;
+            if (!inputReceiver.HasInput())
+                return;
+
+            var readValue = inputReceiver.ReadInput();
+
+            if (p.Item1.IsStatic)
+            {
+                TryInvoke(p.Item1, null, delta, inputReceiver, readValue);
+            }
+            else
+            {
+                var instances =
+                    RunOnInputAttribute.InputClasses.Where(x => x.GetType() == p.Item1.DeclaringType).ToList();
+                foreach (var instance in instances)
+                {
+                    if (!TryInvoke(p.Item1, instance, delta, inputReceiver, readValue))
+                        disposed.Add(instance);
+                }
+            }
+        });
+        disposed.ForEach(p => RunOnInputAttribute.InputClasses.Remove(p));
+    }
+
+    /// <returns>False if the instance was disposed</returns>
+    private static bool TryInvoke(MethodBase method, object instance, float delta, IInputReceiver inputReceiver,
+        object readValue)
+    {
+        try
+        {
+            switch (inputReceiver)
+            {
+                case InputMultiAxis _:
+                case InputAxis _:
+                    method.Invoke(instance, new[] { delta, readValue });
+                    break;
+                case InputTrigger _:
+                case InputReleaseTrigger _:
+                    method.Invoke(instance, Array.Empty<object>());
+                    break;
+                case InputHoldToggle _:
+                case InputBool _:
+                    method.Invoke(instance, new object[] { delta });
+                    break;
+            }
+        }
+        catch (TargetInvocationException ex)
+        {
+            // Disposed object
+            return !(ex.InnerException is ObjectDisposedException);
+        }
+
+        return true;
+    }
+
+    private static void RecalculateAttributes()
+    {
+        foreach (var assembly in AppDomain.CurrentDomain.GetAssemblies())
```

Comment:
Yeah, but what about future mod support? Mod assemblies would not be able to use this class then, if I'm correct?

Reply:
I suppose that's true, but loading a mod assembly wouldn't retrigger this code anyway.
At least blacklist the system binaries by name?

Reply:
meh, I don't like the blacklist solution, because they are too many assemblys.
I could create a whitelist with assemblies to load, which is just the base game at this point.
Future mod managers can add a mod assembly to the list and reload the inputs.